src/transformers/models/bark/modeling_bark.py

```diff
@@ -0,0 +1,1429 @@
+
+        hidden_states = self.transformer.drop(input_embeds + position_embeds)
+        output_shape = input_shape + (hidden_states.size(-1),)
+
+        all_self_attentions = () if output_attentions else None
+        all_hidden_states = () if output_hidden_states else None
+
+        for i, block in enumerate(self.transformer.h):
+            if output_hidden_states:
+                all_hidden_states = all_hidden_states + (hidden_states,)
+
+            outputs = block(
+                hidden_states,
+                attention_mask=attention_mask,
+                head_mask=head_mask[i],
+                output_attentions=output_attentions,
+            )
+
+            hidden_states = outputs[0]
+
+            if output_attentions:
+                all_self_attentions = all_self_attentions + (outputs[1],)
+
+        hidden_states = self.transformer.ln_f(hidden_states)
+        hidden_states = hidden_states.view(output_shape)
+
+        # Add last hidden state
+        if output_hidden_states:
+            all_hidden_states = all_hidden_states + (hidden_states,)
+
+        logits = self.lm_heads[codebook_idx - self.config.n_codes_given](hidden_states)
+
+        loss = None
+        if labels is not None:
+            raise NotImplementedError("Training is not implemented yet")
+
+        if not return_dict:
+            return tuple(v for v in [None, logits, all_hidden_states, all_self_attentions] if v is not None)
+
+        return MaskedLMOutput(
+            loss=loss,
+            logits=logits,
+            hidden_states=all_hidden_states,
+            attentions=all_self_attentions,
+        )
+
+
+@add_start_docstrings(
+    """
+    HugginFace implementation of `Bark`, a text-to-speech model composed of 3 sub-models:
+    1. A semantic model, that refines the input information.
+    2. A coarse acoustics model, that generates a raw version of the audio with the two first codebooks of `Encodec`.
+    3. A fine acoustics model, that refines the raw audio by generating the last 6 codebooks of `Encodec`, following
+       `Bark` original implementation.
+    """,
+    BARK_START_DOCSTRING,
+)
+class BarkModel(BarkPreTrainedModel):
+    # TODO: Add code sample when preprocessing is clearer, and encodec is added.
+    # @add_code_sample_docstrings(
+    #    checkpoint=_CHECKPOINT_FOR_DOC,
+    #    output_type=BaseModelOutputWithPast,
+    #    config_class=_CONFIG_FOR_DOC,
+    # )
+    config_class = BarkConfig
+
+    def __init__(self, config):
+        super().__init__(config)
+
+        self.semantic = BarkSemanticModule(config.semantic_config)
+        self.coarse_acoustics = BarkCoarseAcousticsModule(config.coarse_acoustics_config)
+        self.fine_acoustics = BarkFineAcousticsModule(config.fine_acoustics_config)
+
+        self.codec_model = EncodecModel.from_pretrained(config.pretrained_encodec_name_or_path)
+
+        self.config = config
+
+    def preprocess_histories_before_coarse(
```

Comment:
Should be in `torch` here as well ideally! Is there a reason we can't be? I would only use `numpy` if it's significantly faster (which is usually the case if we do lots of indexing with a small number of indices)

Reply:
No particular reason, I haven't benchmarked it tbh, I've changed to `torch`.